connpicker.go

```diff
@@ -0,0 +1,87 @@
+package gocql
+
+import (
+	"sync"
+	"sync/atomic"
+)
+
+type ConnPicker interface {
+	Pick(token) *Conn
+	Put(*Conn)
+	Remove(conn *Conn)
+	Size() (int, int)
+	Close()
+}
+
+type DefaultConnPicker struct {
+	conns []*Conn
+	pos   uint32
+	size  int
+	mu    sync.RWMutex
+}
+
+func NewDefaultConnPicker(size int) *DefaultConnPicker {
+	return &DefaultConnPicker{
+		size: size,
+	}
+}
+
+func (p *DefaultConnPicker) Remove(conn *Conn) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	for i, candidate := range p.conns {
+		if candidate == conn {
+			p.conns[i] = conn
+			return
+		}
+	}
+}
+
+func (p *DefaultConnPicker) Close() {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	conns := p.conns
+	p.conns = nil
+	for _, conn := range conns {
+		if conn != nil {
+			conn.Close()
+		}
+	}
+}
+
+func (p *DefaultConnPicker) Size() (int, int) {
+	sz := len(p.conns)
+	return sz, p.size - sz
+}
+
+func (p *DefaultConnPicker) Pick(token) *Conn {
+	pos := int(atomic.AddUint32(&p.pos, 1) - 1)
```

Comment:
Not sure the initial reason for this, but would you mind just reading atomically directly?

Reply:
This keeps track of the current position in the conns list. I think it was always like this to step to the next connection in one atomic go.

Reply:
right, ok